src/main/java/com/spikes2212/command/drivetrains/swerve/SwerveModule.java

```diff
@@ -0,0 +1,93 @@
+
+    protected final PIDSettings drivePIDSettings;
+    protected final PIDSettings turnPIDSettings;
+    protected final FeedForwardSettings driveFeedForwardSettings;
+    protected final FeedForwardSettings turnFeedForwardSettings;
+
+    public SwerveModule(String namespaceName, SmartMotorController driveMotor, SmartMotorController turnMotor,
+                        boolean driveInverted, boolean turnInverted, double offset, PIDSettings drivePIDSettings,
+                        PIDSettings turnPIDSettings, FeedForwardSettings driveFeedForwardSettings,
+                        FeedForwardSettings turnFeedForwardSettings) {
+        super(namespaceName);
+        this.driveMotor = driveMotor;
+        this.turnMotor = turnMotor;
+        this.turnInverted = turnInverted;
+        this.driveInverted = driveInverted;
+        this.offset = offset;
+        this.drivePIDSettings = drivePIDSettings;
+        this.turnPIDSettings = turnPIDSettings;
+        this.driveFeedForwardSettings = driveFeedForwardSettings;
+        this.turnFeedForwardSettings = turnFeedForwardSettings;
+        driveMotor.setInverted(driveInverted);
+        turnMotor.setInverted(turnInverted);
+        configureTurnController();
+        configureDriveController();
+        configureAbsoluteEncoder();
+    }
+
+    public void setTargetState(SwerveModuleState targetState, double maxVelocity, boolean usePIDVelocity) {
+        targetState.optimize(Rotation2d.fromDegrees(turnMotor.getPosition()));
+        setTargetAngle(targetState.angle);
+        setTargetVelocity(targetState.speedMetersPerSecond, maxVelocity, usePIDVelocity);
+    }
+
+    public void setTargetAngle(Rotation2d targetAngle) {
+        turnMotor.pidSet(UnifiedControlMode.POSITION, targetAngle.getDegrees(), turnPIDSettings,
+                turnFeedForwardSettings, false);
+    }
+
+
+    public void setTargetVelocity(double targetVelocity, double maxVelocity, boolean usePIDVelocity) {
+        if (usePIDVelocity) {
+            driveMotor.pidSet(UnifiedControlMode.VELOCITY, targetVelocity, drivePIDSettings,
+                    driveFeedForwardSettings, false);
+        } else {
+            driveMotor.set(targetVelocity / maxVelocity);
+        }
+    }
+
+    protected abstract void configureDriveController();
+
+    protected abstract void configureTurnController();
+
+    /**
+        * Defines the contract for configuring an absolute encoder.
+        * Implementations should set direction, offset, and calibration parameters
+        * to align encoder readings with the system’s mechanics.
+     */
+    protected abstract void configureAbsoluteEncoder();
+
+    protected abstract Rotation2d getAbsoluteModuleAngle();
+
+    public void resetRelativeEncoder() {
+        turnMotor.setPosition(getAbsoluteModuleAngle().getDegrees());
+    }
+
+    public SwerveModuleState getState() {
+        return new SwerveModuleState(driveMotor.getVelocity(),
+                getAbsoluteModuleAngle());
```

Comment:
I haven't worked with swerve in a while so correct me if I'm wrong, but I believe that in general you should work with the relative encoder instead of the absolute encoder, because over time they will desync and it's preferred to be less accurate in order to avoid "confusing" the turn controller.